python/ray/data/_internal/pandas_block.py

```diff
@@ -87,7 +92,31 @@ def _table_from_pydict(self, columns: Dict[str, List[Any]]) -> "pandas.DataFrame
 
     def _concat_tables(self, tables: List["pandas.DataFrame"]) -> "pandas.DataFrame":
         pandas = lazy_import_pandas()
-        return pandas.concat(tables, ignore_index=True)
+        from ray.data.extensions.tensor_extension import TensorArray
+
+        df = pandas.concat(tables, ignore_index=True)
+        # Try to convert any ndarray columns to TensorArray columns.
+        # TODO(Clark): Once Pandas supports registering extension types for type
+        # inference on construction, implement as much for NumPy ndarrays and remove
+        # this. See https://github.com/pandas-dev/pandas/issues/41848
+        for col_name, col in df.items():
```

Comment:
@scv119 @ericl @matthewdeng This is an `O(# of columns)` operation that has to be performed when building every Pandas block (i.e. at the end of every UDF), even if the data doesn't contain any tensor columns. For the many-column case (e.g. 10k columns), this is a moderate slowdown, ~400ms on my benchmarking. We need to decide whether this is tenable; we originally didn't add this UX nicety due to this performance concern.

Reply:
curious which operation in the loop is costly? is it `df.items()`, or `df[col_name] = TensorArray(col)`, or something else?

Reply:
It's mostly the Python-level iteration, i.e. consuming the `df.items()` iterator.